GoogleForCreators/web-stories-wp
# Added new copy-paste handlers for both canvas and editor

## Summary

This adds rich-text pasting to the editor. This works both in "pure" canvas and inside edit mode.

## Relevant Technical Choices

- Uses draft-js built-in paste handler
- But removes all irrelevant entities and styling and adds our own limited styling instead
- Uses the same underlying handler for both canvas and edit-mode

## To-do

- [x] Handle pasting in edit-mode
- [x] Handle pasting in canvas
- ~Add karma testing~ (not really possible)

## User-facing changes

Some demos of copying and pasting rich content from various sources - these are all for edit-mode, but it works similarly if not in edit-mode.

#### Google Docs on Mac in Chrome
![rich_paste_docs](https://user-images.githubusercontent.com/637548/97937139-cb919300-1d4b-11eb-9b81-3eb05b8d8db0.gif)

#### Pages on Mac in Chrome
![rich_paste_pages](https://user-images.githubusercontent.com/637548/97937132-c92f3900-1d4b-11eb-86e7-6a808a6e2051.gif)

#### PDF Preview on Mac in Chrome
![rich_paste_pdf](https://user-images.githubusercontent.com/637548/97937145-cfbdb080-1d4b-11eb-8df3-bdb5741ed47e.gif)

#### HTML content on Mac in Chrome
![rich_paste_html](https://user-images.githubusercontent.com/637548/97937143-cd5b5680-1d4b-11eb-8572-e3b0ea02296d.gif)

## Testing Instructions

1. Paste rich-text content
1. Allow for some leeway, as not all formatting works. But some (**bold**, _italic_) should work most of the time, others (lists and <ins>underline</ins>) some of the time. The remaining formatting is never preserved (font, size, color, etc).
1. Test both by pasting when an element is selected or no element is selected (both give the same result, should add a new text element) and when in edit-mode (should paste inside existing text element).

I have tested with [this doc](https://docs.google.com/document/d/12r2hJ4ZqltYKYzF_iQNueKtM5wEaSbiuU_IaFzorgqo/edit?usp=sharing), [this HTML](https://github.com/google/web-stories-wp/issues/4602#issuecomment-720171479), [this PDF](https://drive.google.com/file/d/1W5kkIitrp2kvTrTKi1JlAH9W--7DGHuV/view?usp=sharing) and [this Pages document](https://drive.google.com/file/d/1SkFUQRcM8-VurhlRgrmy9Fehg5kdgDSJ/view?usp=sharing).

---



Fixes #760
